### PR TITLE
Add real oop `finite_difference_jacobian`

### PR DESCRIPTION
**Pull request by huanglangwen**

### CODE REVIEW

File: src/jacobians.jl [hunk 1]
```diff
@@ -28,16 +28,16 @@ function JacobianCache(
         _fx1 = copy(x)
     end
 
-    JacobianCache(x1,_fx,_fx1,fdtype,returntype,inplace;colorvec=colorvec,sparsity=sparsity)
```
Comment (ChrisRackauckas) [inline src/jacobians.jl]: this is upstream breaking?

**Comment by ChrisRackauckas** [comment]
is this non-allocating on static arrays?

**Comment by ChrisRackauckas** [comment]
it would be nice to test the allocs on static arrays.

File: src/jacobians.jl (hunk 1 above)
Comment (huanglangwen) [reply src/jacobians.jl]: Indeed, but it would be meaningless to keep the inplace option.

**Comment by andreasnoack** [comment]
This PR broke the API and therefore downstream pacakges so it shouldn't have been included in a minor release. At a minimum a the method with an `inplace` argument should have been deprecated.

**Pull request merged**